middleware/rewrite/edns0.go

```diff
@@ -0,0 +1,88 @@
+// Package rewrite is middleware for rewriting requests internally to something different.
+package rewrite
+
+import (
+	"encoding/hex"
+	"log"
+	"strconv"
+
+	"github.com/miekg/dns"
+)
+
+// Edns0Rule is a class rewrite rule.
+type Edns0Rule struct {
+	action string
+	code   uint16
+	data   []byte
+	match  []byte // TODO: add match input for replace rule
+}
+
+// SetEDNS0Attrs will alter the request
+func (rule *Edns0Rule) SetEDNS0Attrs(r *dns.Msg) {
+	o := r.IsEdns0()
+	if o == nil {
+		return
+	}
+
+	found := false
+	for _, s := range o.Option {
+		switch e := s.(type) {
+		case *dns.EDNS0_LOCAL:
+			if rule.code == e.Code {
+				if rule.action == "replace" || rule.action == "set" {
+					e.Data = rule.data
+				}
+				found = true
+				break
+			}
+		}
+	}
+
+	// add option if not found
+	if !found && (rule.action == "append" || rule.action == "set") {
+		o.SetDo(true)
+		opt := new(dns.EDNS0_LOCAL)
+		opt.Code = rule.code
+		opt.Data = rule.data
+		o.Option = append(o.Option, opt)
+	}
+
+}
+
+// Initializer
+func (rule Edns0Rule) New(args ...string) Rule {
+	if len(args) < 3 {
+		log.Printf("[WARN] %s is invalid", args)
```

Comment:
I think Bob did that because then he sticks a new empty xyzRule in field.go and doesn't have to change the code in setup.go.